sw/device/lib/otbn.c

```diff
@@ -0,0 +1,162 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "otbn.h"
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/runtime/hart.h"
+
+static bool is_func_ptr(const otbn_t *ctx, const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space >= ctx->app->imem_start &&
+         ptr->ptr_in_cpu_addr_space <= ctx->app->imem_end;
+}
+
+static uint32_t func_ptr_in_otbn_addr_space(const otbn_t *ctx,
+                                            const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space - ctx->app->imem_start;
+}
+
+static bool is_data_ptr(const otbn_t *ctx, const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space >= ctx->app->dmem_start &&
+         ptr->ptr_in_cpu_addr_space <= ctx->app->dmem_end;
+}
+
+static uint32_t data_ptr_in_otbn_addr_space(const otbn_t *ctx,
+                                            const otbn_ptr_t *ptr) {
+  return ptr->ptr_in_cpu_addr_space - ctx->app->dmem_start;
+}
+
+otbn_result_t otbn_busy_wait_for_done(otbn_t *ctx) {
+  bool busy = true;
+  while (busy) {
+    if (dif_otbn_is_busy(&ctx->dif, &busy) != kDifOtbnOk) {
+      return kOtbnError;
+    }
+  }
+
+  dif_otbn_err_code_t err_code;
+  if (dif_otbn_get_err_code(&ctx->dif, &err_code) != kDifOtbnOk) {
+    return kOtbnError;
+  }
+  if (err_code != kDifOtbnErrCodeNoError) {
+    return kOtbnExecutionFailed;
+  }
+  return kOtbnOk;
+}
+
+otbn_result_t otbn_init(otbn_t *ctx, const dif_otbn_config_t *dif_config) {
+  if (ctx == NULL || dif_config == NULL) {
+    return kOtbnBadArg;
+  }
+
+  ctx->app = NULL;
+  ctx->app_is_loaded = false;
+
+  if (dif_otbn_init(dif_config, &ctx->dif) != kDifOtbnOk) {
+    return kOtbnError;
+  }
+
+  return kOtbnOk;
+}
+
+otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t *app) {
+  if (app->imem_end <= app->imem_start || app->dmem_end < app->dmem_start) {
+    return kOtbnBadArg;
+  }
+
+  const size_t imem_size = app->imem_end - app->imem_start;
+  const size_t dmem_size = app->dmem_end - app->dmem_start;
+
+  // Instruction and data memory images must be multiples of 32b words.
+  if (imem_size % 4 != 0 || dmem_size % 4 != 0) {
```

Comment:
sizeof(uint32_t), maybe?

Reply:
I don't think that makes the code more readable or adds other value.

Reply:
Changed to use `sizeof(uint32_t)`.